### DISCUSSION TITLE
Unable to track changes from async constructors

### DISCUSSION THREAD

**Discussion by WesleyYue** [pmndrs/valtio#830] [Bug report]
It seems that the proxy is not able to track changes from async constructors. I have some intuition as to why that is, but I am wondering if there's something the library can implement to support this?

[CodeSandbox repro](https://codesandbox.io/p/devbox/intelligent-margulis-kdv9cm?file=%2Fsrc%2Fvaltio.test.ts%3A1%2C1-23%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clpkt96tu00063b6h9ak4f8di%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clpkt96tt00023b6h6nvv7la2%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clpkt96tt00043b6h30vls1g9%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clpkt96tt00053b6hkqjlticf%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B60%252C40%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clpkt96tt00023b6h6nvv7la2%2522%253A%257B%2522id%2522%253A%2522clpkt96tt00023b6h6nvv7la2%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpktf71g00023b6h6ysaj9dn%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A1%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A23%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fvaltio.test.ts%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clpktf71g00023b6h6ysaj9dn%2522%257D%252C%2522clpkt96tt00053b6hkqjlticf%2522%253A%257B%2522id%2522%253A%2522clpkt96tt00053b6hkqjlticf%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clpkt96tt00043b6h30vls1g9%2522%253A%257B%2522id%2522%253A%2522clpkt96tt00043b6h30vls1g9%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpkt96tt00033b6hytqj32vs%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clpkt96xe0002efhkgyp12uv5%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clpktb2b2006z3b6he469fxir%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clpkt96tt00033b6hytqj32vs%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by dai-shi** [comment]
It works as expected. `proxy(initialObject)` returns a new proxy object based on `initialObject`. In your first test, `this` points to `initialObject`, which is not the proxy object. So, its operation isn't tracked.
